### DISCUSSION TITLE
Structular regular expressions for multiple cursors

### DISCUSSION THREAD

**Discussion by gavlooth** [helix-editor/helix#7292] [Ideas] [answered]
There is a vim clone - vis - that has multiple cursor support and a very neat future that pairs very well. Essentially, you create complex operations by combining sam9's regular expressions. 
This might be a killer future.

**Comment by kirawi** [comment]
It's not obvious to me what structural regex is, but the name alone suggests something along the lines of: https://github.com/helix-editor/helix/discussions/3476

**Reply by gavlooth** [reply]
It's an alternative way to operate on selections by chaining multiple regex commands. For me is much more convenient than
memorizing the selection commands.
There is this cli tool https://github.com/zyedidia/sregx that I use with neovim
I might be able to use it in a similar way with helix through the :pipe command, but i think it could also be an awesome future for the
core and very easy to implement.
It doesnt seem to be similar with #3476    since the issue focuses only on queries (if i understand it correctly)

**Reply by gavlooth** [reply]
The original paper is here: http://doc.cat-v.org/bell_labs/structural_regexps/se.pdf
And you can see how powerful is with multiple cursors here : https://asciinema.org/a/41361
I am not the author or anything

**Reply by pascalkuthe** [reply]
We already have regex based selection: `s` and `K`

**Reply by gavlooth** [reply]
@pascalkuthe  sregex is not regex based selection. Is chaining commands based on regexes on selections eg 
"Change all occurrences of the complete word "foo" to "bar" except those occurring in double or single quoted strings:

`y/".*"/ y/'.*'/ x/[a-zA-Z]+/ g/^foo$/ c/bar/ | p"`

`x/<p>/<cmd>:` returns a string where all substrings matching the regular expression <p> have been replaced with the return value of <cmd> applied to the particular substring.
`y/<p>/<cmd>: `returns a string where each part of the string that is not matched by <p> is replaced by applying <cmd> to the particular unmatched string.
... 
The idea is that you select something and then chain multiple regex commands at once to manipulate text

**Reply by gavlooth** [reply]
I am not saying that you cannot do selection manipulation with  key chords, but it is  convenient to have the
command language.

**Reply by pascalkuthe** [reply, marked as answer]
that's an explicit non-goal for helix. That's just  vim search/replace (maybe more powerful). Helix is fundamentally based around selections. A command either manipulates selection or changes content. Fused commands purposefully don't exist and won't be added. That's a core design choice
